Configure Envoy alpn_protocols based on service protocol

oulman opened the pull request:
### Description

This PR is largely based on @blake's work in https://github.com/hashicorp/consul/pull/12170 but updated to use the TLSContext functions introduced in https://github.com/hashicorp/consul/pull/13321. I'm not sure what the best path forward for configuring this from the Consul teams perspective but I wanted to raise a PR to start a conversation.

This fixes https://github.com/hashicorp/consul/issues/11907, resolves https://github.com/hashicorp/consul/issues/12106, and [this internal feature request](https://go.hashi.co/alpn_protocols_fr).

### Testing & Reproduction steps

In my testing I verified that gRPC and HTTP2 services can negotiate h2 and http/1.1 and that services configured as HTTP will only negotiate http/1.1. 

### Links
Include any links here that might be helpful for people reviewing your PR (Tickets, GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.

Please be mindful not to leak any customer or confidential information. HashiCorp employees may want to use our internal URL shortener to obfuscate links.

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] not a security concern

malizz commented on agent/xds/listeners.go:
```diff
@@ -1022,6 +1023,19 @@ func (s *ResourceGenerator) injectConnectTLSForPublicListener(cfgSnap *proxycfg.
+	var alpnProtocols []string
+
+	switch protocol {
+	case "grpc", "http2":
```
I believe `grpc` and `http2` cases behave differently. `http2` should only return `h2`.

oulman replied:
Thanks! I've updated the function to return `h2` only for `http2`.

malizz commented:
Hi @oulman! Thanks for this PR, it looks good, but we can work on adding more tests and verifying the intended behavior.

oulman commented:
Hi @malizz - Thanks! I've pushed a couple of changes.

- The linter was failing because the error handling on ParseProxyConfig was empty. The original PR had a commented out call to a logger but I didn't see one available in the function. In the interim I went ahead and excluded that [here](https://github.com/hashicorp/consul/pull/14356/commits/b4cb2537014cf20ab52b9d037f6e7de3ba180bae)
- I created a [test for getAlpnProtocol()](https://github.com/hashicorp/consul/pull/14356/commits/e2e342bd9b10f2bdb8f0ffa9b1c0754ac5491ab4)
- I added [tests](https://github.com/hashicorp/consul/pull/14356/commits/bd6ec7f1ce510462853bed071a477d9a2fb8ed42) for http2 and grpc public listeners to validate that the Envoy `alpn_protocols` setting was getting configured correctly.

malizz commented:
@oulman Question: you made the changes to listeners tlsContext only, but I’m wondering if we need to update the TLS contexts for other listeners (ingress listener for example) or pass through clusters too.

oulman commented:
Hi @malizz! I updated the context for the Ingress Gateway listener and added tests.

Looking at pass-through clusters I'm a little confused on what actually needs to be done to support this. I found some issues  ([#1](https://github.com/envoyproxy/envoy/issues/11909#issuecomment-654903211), [#2](https://github.com/envoyproxy/envoy/issues/14378#issuecomment-743713707)) that make it sound like for Envoy <-> Envoy traffic ALPN isn't used in protocol negotiation. If mesh-to-mesh is OK as-is, do we need to enable ALPN negotiation for external upstream HTTP clusters?[1]

Thanks!

1. I also found [this Envoy PR](https://github.com/envoyproxy/envoy/pull/13922) which enables ALPN negotiation via [http_protocol_options](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/upstreams/http/v3/http_protocol_options.proto#envoy-v3-api-msg-extensions-upstreams-http-v3-httpprotocoloptions-autohttpconfig).

malizz commented on agent/proxycfg/testing_ingress_gateway.go:
```diff
@@ -1485,6 +1485,164 @@ func TestConfigSnapshotIngressGateway_SingleTLSListener(t testing.T) *ConfigSnap
 		})
 }
 
```
I have a question, is it possible to have an Ingress gateway listener that listens to port x for grpc and port y for http2?

blake replied:
A listener only binds to a single port. It is possible to configure an ingress gateway with one listener on port X with the protocol as `grpc`, and different listener on port Y with the protocol configured as `http2`.

malizz replied:
Thanks @blake. @oulman could you add a test function for this use case?

oulman replied:
Sure! I just added a test for `TestConfigSnapshotIngressGateway_MultiTLSListener_MixedHTTP2gRPC` which validates two listeners (grpc and http2) with per-listener TLSConfig's.

The other scenario this made me think about is configuring TLS at the gateway [level](https://www.consul.io/docs/connect/config-entries/ingress-gateway#tls). From my read it looks like the Gateway level TLS is still merged down to the listener level and setting ALPN in `makeDownstreamTLSContextFromSnapshotListenerConfig()` still applies. I added another test `TestConfigSnapshotIngressGateway_GWTLSListener_MixedHTTP2gRPC` to validate that case. 

Thanks!

malizz reviewed (changes requested):
The PR looks mostly great. I added a question and requested a few changes. Thanks.

oulman commented on agent/xds/listeners_test.go:
```diff
@@ -1115,3 +1141,34 @@ func TestResolveListenerSDSConfig(t *testing.T) {
+			protocol: "",
+			want:     nil,
+		},
+	}
```
👍 I added an explicit case for `tcp`.

oulman commented on agent/xds/listeners.go:
```diff
@@ -1030,13 +1046,22 @@ func createDownstreamTransportSocketForConnectTLS(cfgSnap *proxycfg.ConfigSnapsh
+	// Determine listener protocol type from configured service protocol. Don't hard fail on a config typo,
+	//The parse func returns default config if there is an error, so it's safe to continue.
+	cfg, _ := ParseProxyConfig(cfgSnap.Proxy.Config)
+
```
🤔 Right now calling getAlpnProtocols with `tcp` is returning an implicit nil because it's not matched. To confirm, we want to add an case statement for tcp but it should still return the empty `alpnProtocols` slice.

malizz reviewed (approved):
LGTM, thanks!

malizz commented on agent/xds/listeners.go:
```diff
@@ -1022,6 +1023,21 @@ func (s *ResourceGenerator) injectConnectTLSForPublicListener(cfgSnap *proxycfg.
+	switch protocol {
+	case "grpc":
+		alpnProtocols = append(alpnProtocols, "h2", "http/1.1")
+	case "http2":
```
I think the correct list of protocols is "h2" and "http/1.1" here. Sorry for the mixup.

oulman replied:
OK. Thanks! I pushed a new commit that switches this.
